wagtail/core/tests/test_collectors.py

```diff
@@ -0,0 +1,312 @@
+            self.assertListEqual(uses, [(self.obj6, self.image6),
+                                        (self.obj6, self.image10),
+                                        (self.obj6, self.image7),
+                                        (self.obj6, self.image8),
+                                        (self.obj6, self.image9),
+                                        (self.obj7, self.image9),
+                                        (self.obj8, self.image8),
+                                        (self.obj9, self.image7),
+                                        (self.obj10, self.image6)])
+
+    def test_rich_text_block(self):
+        with self.assertNumQueries(2):
+            uses = list(self.collector.find_objects(self.image11))
+            self.assertListEqual(uses, [(self.obj11, self.image11)])
+
+
+class GetAllUsesTest(TestCase):
+    def setUp(self):
+        i = 0
+
+        def create_page(body):
+            nonlocal i
+            home = Page.objects.get(depth=2)
+            page = DefaultRichTextFieldPage(
+                title='%s' % i, body=body, slug='%s' % i)
+            home.add_child(instance=page)
+            i += 1
+            return page
+
+        self.obj0 = create_page('<p>Nothing worth of interest.</p>')
+        self.obj1 = create_page('<p>Nothing worth of interest.</p>')
+        self.obj2 = PageChooserModel.objects.create(page=self.obj1)
+        self.obj3 = create_page('<p>Nothing worth of interest.</p>')
+        self.obj4 = create_page('<p><a linktype="page" id="%s">3</a></p>'
+                                % self.obj3.pk)
+        self.obj5 = Image.objects.create(title='Test image 1',
+                                         file=get_test_image_file())
+        self.obj6 = StreamModel.objects.create(body=[('image', self.obj5)])
+        self.obj7 = Image.objects.create(title='Test image',
+                                         file=get_test_image_file())
+        self.obj8 = self.obj7.get_rendition('original')
+        self.obj9 = StreamModel.objects.create(body=[('image', self.obj7)])
+        self.obj10 = create_page(
+            '<p><embed alt="bodyline" embedtype="image" '
+            'format="centered" id="%s"/></p>' % self.obj7.pk)
+
+    def test_empty(self):
+        with self.assertNumQueries(53):
+            uses = list(get_all_uses(self.obj0))
+            self.assertListEqual(uses, [])
+
+    def test_foreign_key(self):
+        with self.assertNumQueries(53):
+            uses = list(get_all_uses(self.obj1))
+            self.assertListEqual(uses, [self.obj2])
+
+    def test_rich_text(self):
+        with self.assertNumQueries(55):
+            uses = list(get_all_uses(self.obj3))
+            self.assertListEqual(uses, [self.obj4])
+
+    def test_streamfield(self):
+        with self.assertNumQueries(55):
+            uses = list(get_all_uses(self.obj3))
+            self.assertListEqual(uses, [self.obj4])
+
+    def test_multiple(self):
```

Comment:
I ran into a tricky issue with this test. If you run only these new test cases (`wagtail.core.tests.test_collectors`) without also running the full test suite, this specific test fails with `AssertionError: 47 != 46 : 47 queries executed, 46 expected`.

The extra 47th query that gets run if you **don't** first run one of those test cases is

```sql
18. SELECT "django_content_type"."id", "django_content_type"."app_label", "django_content_type"."model" FROM "django_content_type" WHERE ("django_content_type"."app_label" = 'wagtailimages' AND "django_content_type"."model" = 'image')
```

So this appears to be something where the image content type maybe [isn't cached](https://docs.djangoproject.com/en/2.0/ref/contrib/contenttypes/#django.contrib.contenttypes.models.ContentTypeManager.clear_cache) as part of the Django content type framework.

I dug around and one way to get `wagtail.core.tests.test_collectors.GetAllUsesTest.test_multiple` to pass is if you first run a `wagtail.snippets.tests` test case that loads [the `test.json` fixture](https://github.com/wagtail/wagtail/blob/db6d36845f3f2c5d7009a22421c2efab9968aa24/wagtail/tests/testapp/fixtures/test.json) (for example [`wagtail.snippets.tests.TestSnippetChooserBlock`](https://github.com/wagtail/wagtail/blob/db6d36845f3f2c5d7009a22421c2efab9968aa24/wagtail/snippets/tests.py#L7199)). I'm not super familiar with how content types work, but presumably there's some step in this other test that causes that content type to get cached.

One option would be to do [something like this](https://github.com/django/django/commit/7cd3f1c29595d1da7f37d29e7c3bc6a7a314cd1d) to manually clear the cache first (suggested by [this Django ticket](https://code.djangoproject.com/ticket/23746)) to avoid this problem. Ideally all tests should be runnable on their own; is there a more flexible approach than using a hardcoded number of SQL queries here?  If and when new models get added to Wagtail anywhere, these tests will have to get updated as the number of queries increase.

Reply:
You’re totally right, I’ll change the tests in order to clear the ContentType cache before.
That way, we’ll test the maximum amount of queries instead of the minimum.